accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -1884,27 +1884,51 @@ var [int a, b] = <num>[1, 2];   // List<num>.
 
 To type check a pattern `p` being matched against a value of type `M`:
 
-*   **Logical-or** and **logical-and**: Type check each branch using `M` as the
-    matched value type.
+*   **Logical-or**: Type check each branch using `M` as the
+    matched value type. The required type of the pattern is `Object?`.
+    *`M` will be used to perform checks on each operand, whose required types
+    may be more strict.*
+
+*   **Logical-and**: Type check each branch using `M` as the matched value
+    type of the left operand, and using the promoted type of the
+    scrutinee from the left operand as the matched value type of the right hand
+    operand *(that's `M` again, if there is no promotion)*. The required type of
+    the pattern is `Object?`.
+    *The chosen matched value type will be used to perform checks on each
+    operand, whose required types may be more strict.*
+
+*   **Relational**: Consider the relational pattern `op c` where `op` is one
+    of the following operators: `==`, `!=`, `<`, `<=`, `>=`, `>`, and `c` is an
+    expression.
 
-*   **Relational**:
+    If `M` is `dynamic`: Type check `c` in context `_`; an error occurs
+    if `c` is not a constant expression; no further checks are
+    performed. Otherwise *(when `M` is not `dynamic`)*:
+
+    1.  A compile-time error occurs if `M` does not have an operator `op`,
+        and there is no available and applicable extension operator `op`. 
+        Let `A` be the type of the formal parameter of the given operator
+        declaration, and let `R` be the return type.
+
+    2.  A compile-time error occurs if `R` is not assignable to `bool`.
 
-    1.  Let `C` be the static type of the right operand constant expression.
+    3.  Type check `c` with context type `A`. A compile-time error occurs if
+        `c` is not a constant expression. Let `C` be the static type of `c`.
 
-    2.  If the operator is a comparison (`<`, `<=`, `>`, or `>=`), then it is a
-        compile-time error if:
+    4.  If `op` is `==` or `!=` then a compile-time error occurs if `C` is not
+        assignable to `A?`. Otherwise `op` is `<`, `<=`, `>=`, or `>`, and a
+        compile-time error occurs if `C` is not assignable to `A`.
 
-        *   `M` does not define that operator,
-        *   `C` is not assignable to the operator's parameter type,
-        *   or if the operator's return type is not assignable to `bool`.
+    *The language screens out `null` before calling the underlying `==`
+    method, which is why `T?` is the allowed type for equality checks. Since
```

Comment:
`T?` -> `C?`

There is no `T` definition in scope here, but we use `C?` above.